### ISSUE TITLE
Clang accepts invalid program invovling function parameter as non-type template argument

### ISSUE THREAD

**Issue by ranaanoop** [llvm/llvm-project#83101]
The following program is accepted by clang but rejected by both gcc and clang. [Demo](https://godbolt.org/z/M61hq7zv4):

```
#include <array>

auto function(auto s) -> std::array<int, s + s> {
    return {};
}

int main() {
    auto const a = function(std::integral_constant<int, 3>{});
    return a.size();
}

```

**Comment by ranaanoop** [comment]
Similarly, clang also accepts:

```
#include <array>

auto function(auto s) -> std::array<int, s> {
    return {};
}

int main() {
    auto const a = function(std::integral_constant<int, 3>{});
    return a.size();
}

```

**Comment by shafik** [comment]
Clang looks wrong here, we are trying to use the value of the function parameter and it should not be odr-useable in this case. I think [basic.def.odr p10](https://eel.is/c++draft/basic.def.odr#10) covers this but it is a bit wordy.

CC @zygoloid
